README.md
# *date-pt-br*

[![NPM Version](https://badge.fury.io/js/date-pt-br.svg?style=flat)](https://npmjs.org/package/date-pt-br)
[![Build Status](https://travis-ci.org/victorgianvechio/date-pt-br.svg?branch=master)](https://travis-ci.org/victorgianvechio/date-pt-br)
[![license](https://img.shields.io/npm/l/date-pt-br.svg)](https://github.com/victorgianvechio/date-pt-br/blob/master/LICENSE)
[![npm](https://img.shields.io/npm/dt/date-pt-br.svg)](https://www.npmjs.com/package/date-pt-br)
![monthly npm downloads for date-pt-br](https://img.shields.io/npm/dm/date-pt-br.svg)
[![Dependency Status](https://david-dm.org/victorgianvechio/date-pt-br.svg)](https://david-dm.org/victorgianvechio/date-pt-br)


Get current date in pt-BR or format other dates to pt-BR.

**Table of content**

-   [Installation](#installation)
    -   [Scripts](#scripts)
    -   [Usage](#usage)
-   [Changelog](#changelog)
-   [Config Functions](#config-functions)
    -   [config(object)](#configobject)
    -   [setDefaultConfig()](#setdefaultconfig)
-   [Date Functions](#date-functions)
    -   [getDay(date)](#getdaydate)
    -   [getMonthNumber(date)](#getmonthnumberdate)
    -   [getMonth(date)](#getmonthdate)
    -   [getShortMonth(date)](#getshortmonthdate)
    -   [getYear(date)](#getyeardate)
    -   [getShortYear(date)](#getshortyeardate)
    -   [getWeekdayNumber(date)](#getweekdaynumberdate)
    -   [getWeekday(date)](#getweekdaydate)
    -   [getDate(date)](#getdatedate)
    -   [getExtendedDate(date)](#getextendeddatedate)
    -   [getExtendedWeekdayDate(date)](#getextendedweekdaydatedate)
    -   [getMonthNumberYear(date)](#getmonthnumberyeardate)
    -   [getMonthYear(date)](#getmonthyeardate)
    -   [getShortMonthYear(date)](#getshortmonthyeardate)
    -   [getShortMonthShortYear(date)](#getshortmonthshortyeardate)
    -   [getHour(date)](#gethourdate)
    -   [getExtendedHour(date)](#getextendedhourdate)
    -   [getMinute(date)](#getminutedate)
    -   [getExtendedMinute(date)](#getextendedminutedate)
    -   [getSecond(date)](#getseconddate)
    -   [getExtendedSecond(date)](#getextendedseconddate)
    -   [getTime(date)](#gettimedate)
    -   [getExtendedTime(date)](#getextendedtimedate)
    -   [getHourMinute(date)](#gethourminutedate)
    -   [getExtendedHourMinute(date)](#getextendedhourminutedate)
    -   [getDateTime(date)](#getdatetimedate)
    -   [getExtendedDateTime(date)](#getextendeddatetimedate)
    -   [getExtendedCityDate(date)](#getextendedcitydatedate)
    -   [getExtendedRegionDate(date)](#getextendedregiondatedate)
    -   [getExtendedShortRegionDate(date)](#getextendedshortregiondatedate)
    -   [getExtendedCountryDate(date)](#getextendedcountrydatedate)
    -   [getExtendedShortCountryDate(date)](#getextendedshortcountrydatedate)
-   [Current Locale Functions](#current-locale-functions)
    -   [getCity()](#getcity)
    -   [getRegion()](#getregion)
    -   [getShortRegion()](#getshortregion)
    -   [getCountry()](#getcountry)
    -   [getShortCountry()](#getshortcountry)
    -   [getTimezone()](#gettimezone)
    -   [getZipcode()](#getzipcode)
-   [License](#license)

---

## Installation

via npm:

```sh
npm i -S date-pt-br
```

### Scripts

list all functions:

```sh
npm run list-functions
```

### Usage

```javascript
const DatePtBR = require('date-pt-br')
const date = new DatePtBR()
```

## Changelog

see the update notes at [CHANGELOG](https://github.com/victorgianvechio/date-pt-br/blob/master/CHANGELOG.md).

## Config Functions

### config(object)

delimiters and sep are used by the functions.

-   {string} **delimiterDate** - used to separate day, month and year

    -   default: **' / '**
    -   e.g.: 02 **/** 11 **/** 2019

-   {string} **delimiterTime** - used to separate hours, minutes and seconds

    -   default: **' : '**
    -   e.g.: 08 **:** 53 **:** 20

-   {string} **sepDateTime** - used to separate date and time

    -   default: **' , '**
    -   e.g.: 02/01/2019 **,** 08:53:20

-   {string} **sepDate** - used to separate date

    -   default: **' de '**
    -   e.g.: 02 **de** Janeiro **de** 2019

-   {string} **sepTime** - used to separate time

    -   default: **' e '**
    -   e.g.: 08 horas **e** 53 minutos **e** 20 segundos

```javascript
const DatePtBR = require('date-pt-br')
const date = new DatePtBR()

date.config({
    delimiterDate: '/',
    delimiterTime: ':',
    sepDateTime: ', ',
    sepDate: ' de ',
    sepTime: ' e '
})
```

### setDefaultConfig()

```javascript
date.setDefaultConfig()

// delimiterDate: '/',
// delimiterTime: ':',
// sepDateTime: ', ',
// sepDate: ' de ',
// sepTime: ' e '
```

## Date Functions

All functions below receive an **optional** date as a parameter. If no date is passed the function uses the **current** date.

-   {string} **date** - accepts various formats _(optional)_

**e.g.**:

```javascript
date.getDateTime('2012-01-26T13:51:50.417-07:00') // => 26/01/2012, 18:51:50

date.getTime('05/07/2019 08:54:32') // => 08:54:32

date.getDate('October 15, 1996 05: 35: 32') // => 15/10/1996

date.getExtendedWeekdayDate('2017-09-08T15:25:53Z') // => Sexta-Feira, 08 de Setembro de 2017

date.getMonthNumber('05 October 2011 14:48 UTC') // => 10

date.getDateTime('07/24/2015') // => 24/07/2018, 00:00:00

date.getWeekday('05 October 2011 14:48 UTC') // => Quarta-Feira

date.getExtendedTime('Wed Oct 05 2011 16:48:00 GMT+0200 (CEST)' // => 11 horas e 48 minutos e 00 segundos

date.getShortMonthYear('2011-10-05T14:48:00.000Z') // => Out/2011

date.getDay('24HJERH87LK4') // => Invalid Date
```

If the date passed is invalid the functions will return **'Invalid Date'**.

### getDay(date)

-   returns day **(01 a 31)**

```javascript
let val = date.getDay() // => 27
```

### getMonthNumber(date)

-   returns month number **(01 a 12)**

```javascript
let val = date.getMonthNumber() // => 06
```

### getMonth(date)

```javascript
let val = date.getMonth() // => Junho
```

| number | month     |
| :----- | :-------- |
| 01     | Janeiro   |
| 02     | Fevereiro |
| 03     | Março     |
| 04     | Abril     |
| 05     | Maio      |
| 06     | Junho     |
| 07     | Julho     |
| 08     | Agosto    |
| 09     | Setembro  |
| 10     | Outubro   |
| 11     | Novembro  |
| 12     | Dezembro  |

### getShortMonth(date)

```javascript
let val = date.getShortMonth() // => Jun
```

| number | month |
| :----- | :-----|
| 01     | Jan   |
| 02     | Fev   |
| 03     | Mar   |
| 04     | Abr   |
| 05     | Mai   |
| 06     | Jun   |
| 07     | Jul   |
| 08     | Ago   |
| 09     | Set   |
| 10     | Out   |
| 11     | Nov   |
| 12     | Dez   |

### getYear(date)

```javascript
let val = date.getYear() // => 2019
```

### getShortYear(date)

```javascript
let val = date.getShortYear('07/18/2019') // => 19
```

### getWeekdayNumber(date)

-   returns the number of the week **(1 a 7)**

```javascript
let val = date.getWeekdayNumber() // => 5
```

### getWeekday(date)

```javascript
let val = date.getWeekday() // => Quinta-Feira
```

| number | weekday       |
| :----- | :------------ |
| 1      | Domingo       |
| 2      | Segunda-Feira |
| 3      | Terça-Feira   |
| 4      | Quarta-Feira  |
| 5      | Quinta-Feira  |
| 6      | Sexta-Feira   |
| 7      | Sábado        |

### getDate(date)

```javascript
let val = date.getDate() // => 27/06/2019
```

### getExtendedDate(date)

```javascript
let val = date.getExtendedDate() // => 27 de Junho de 2019
```

### getExtendedWeekdayDate(date)

```javascript
let val = date.getExtendedWeekdayDate() // => Quinta-Feira, 27 de Junho de 2019
```

### getMonthNumberYear(date)

```javascript
let val = date.getMonthNumberYear() // => 06/2019
```

### getMonthYear(date)

```javascript
let val = date.getMonthYear() // => Junho/2019
```

### getShortMonthYear(date)

```javascript
let val = date.getShortMonthYear() // => Jun/2019
```

### getShortMonthShortYear(date)

```javascript
let val = date.getShortMonthYear() // => Jun/19
```

### getHour(date)

-   return hours **(01 a 23)**

```javascript
let val = date.getHour() // => 08
```

### getExtendedHour(date)

```javascript
let val = date.getExtendedHour() // => 08 horas
```

### getMinute(date)

-   return minutes **(01 a 59)**

```javascript
let val = date.getMinute() // => 53
```

### getExtendedMinute(date)

```javascript
let val = date.getExtendedMinute() // => 53 minutos
```

### getSecond(date)

-   returns seconds **(01 a 59)**

```javascript
let val = date.getSecond() // => 32
```

### getExtendedSecond(date)

```javascript
let val = date.getExtendedSecond() // => 32 segundos
```

### getTime(date)

```javascript
let val = date.getTime() // => 08:53:32
```

### getExtendedTime(date)

```javascript
let val = date.getExtendedTime() // => 08 horas e 53 minutos e 32 segundos
```

### getHourMinute(date)

```javascript
let val = date.getHourMinute() // => 08:53
```

### getExtendedHourMinute(date)

```javascript
let val = date.getExtendedHourMinute() // => 08 horas e 53 minutos
```

### getDateTime(date)

```javascript
let val = date.getDateTime() // => 27/06/2019 08:53:32
```

### getExtendedDateTime(date)

```javascript
let val = date.getExtendedDateTime() // => 27 de Junho de 2019, 08 horas e 53 minutos e 32 segundos
```

### getExtendedCityDate(date)

_asynchronous function_

```javascript
let val = await date.getExtendedCityDate() // =>  Marília, 22 de Julho de 2019
```

### getExtendedRegionDate(date)

_asynchronous function_

```javascript
let val = await date.getExtendedRegionDate() // =>  São Paulo, 22 de Julho de 2019
```

### getExtendedShortRegionDate(date)

_asynchronous function_

```javascript
let val = await date.getExtendedShortRegionDate() // =>  SP, 22 de Julho de 2019
```

### getExtendedCountryDate(date)

_asynchronous function_

```javascript
let val = await date.getExtendedCountryDate() // =>  Brazil, 22 de Julho de 2019
```

### getExtendedShortCountryDate(date)

_asynchronous function_

```javascript
let val = await date.getExtendedShortCountryDate() // =>  BR, 22 de Julho de 2019
```

## Current Locale Functions

All functions bellow are _asynchronous_ and returns current locale informations.

### getCity()

_asynchronous function_

```javascript
let val = await date.getCity() // => Marília
```

### getRegion()

_asynchronous function_

```javascript
let val = await date.getRegion() // => São Paulo
```

### getShortRegion()

_asynchronous function_

```javascript
let val = await date.getShortRegion() // => SP
```

### getCountry()

_asynchronous function_

```javascript
let val = await date.getCountry() // => Brazil
```

### getShortCountry()

_asynchronous function_

```javascript
let val = await date.getShortCountry() // => BR
```

### getTimezone()

_asynchronous function_

```javascript
let val = await date.getTimezone() // => America/Sao_Paulo
```

### getZipcode()

_asynchronous function_

```javascript
let val = await date.getZipcode() // => 17350-000
```

## License

MIT License

Copyright ® 2019 Victor Gianvechio

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
# Change Log

## date-pt-br v1.2.2 (22 Jul 2019)

-   Fixed README;

### New functions

All new functions is _asynchronous_

-   date functions:
     -   _[getExtendedCityDate(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getextendedcitydatedate)_
    -   _[getExtendedRegionDate(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getextendedregiondatedate)_
     -   _[getExtendedShortRegionDate(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getextendedshortregiondatedate)_
    -   _[getExtendedCountry(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getextendedcountrydatedate)_
    -   _[getExtendedShortCountry(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getextendedshortcountrydatedate)_

-   current locale functions:

    -   _[getCity()](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getcity)_
    -   _[getRegion()](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getregion)_
    -   _[getShortRegion()](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getshortregion)_
    -   _[getCountry()](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getcountry)_
    -   _[getShortCountry()](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getshortcountry)_
    -   _[getTimezone()](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#gettimezone)_
    -   _[getZipcode()](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getzipcode)_

## date-pt-br v1.2.1 (18 Jul 2019)

-   Fixed some bugs;

-   New functions:
    -   _[getShortYear(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getshortyeardate)_
    -   _[getShortMonthShortYear(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getshortmonthshortyeardate)_

## date-pt-br v1.2.0 (17 Jul 2019)

### Huge Updates

-   All codes were refactored;

-   Now date-pt-br is a **class** and needs to be **instantiated**;

-   All functions have only one **optional** parameter **(date)**. If no date is passed the function will use the current date;

-   Delimiters and sep must be configured after instance;

e.g.:

```javascript
const DatePtBR = require('date-pt-br')
const date = new DatePtBR()

date.config({
    delimiterDate: '/',
    delimiterTime: ':',
    sepDateTime: ', ',
    sepDate: ' de ',
    sepTime: ' e '
})

date.getDay() // => 17
date.getDay('07/23/2019') // 23
```

#### Removed Functions

-   _getExtendedCityDate(city)_;
-   _toPtBr(date, delimiter)_

it does not make sense since all the functions already do the conversion after this update;

#### New functions

-   _[getShortMonth(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getshortmonthdate)_;
-   _[getShortMonthYear(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getshortmonthyeardate)_;
-   _[getExtendedDateTime(date)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#getextendeddatetimedate)_;
-   _[config(object)](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#configobject)_;
-   _[setDefaultConfig()](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md#setdefaultconfig)_;

see more about this update at [README](https://github.com/victorgianvechio/date-pt-br/blob/master/README.md).

## date-pt-br v1.1.0 (10 Jul 2019)

-   New function _toPtBr(date, delimiter)_ that convert date to pt-BR;

## date-pt-br v1.0.9 (02 Jul 2019)

-   New function _getDateTime(delimiter1, sep, delimiter2)_ that returns date and time;

## date-pt-br v1.0.8 (01 Jul 2019)

-   Fixed _getDay()_ function that returns '1' instead of '01';
-   Fixed some rules in _.eslint_;

## date-pt-br v1.0.7 (27 Jun 2019)

-   Adjusted README to make it easier to use;

-   Refactored code;

-   Added the _optional_ parameter _delimiter_ in functions:
    -   getTime(_delimiter_)
    -   getExtendedTime(_delimiter_)
    -   getHourMinute(_delimiter_)
    -   getExtendedHourMinute(_delimiter_)

## date-pt-br v1.0.6 (26 Jun 2019)

-   README and code style;

## date-pt-br v1.0.5 (26 Jun 2019)

-   Added script:

```sh
npm run list-functions
```


scripts/functions.js
const DatePtBR = require('../lib/date-pt-br.js')
const date = new DatePtBR()
;(async () => {
    console.log(`\ngetDay() =>`, date.getDay())
    console.log(`\ngetMonthNumber() =>`, date.getMonthNumber())
    console.log(`\ngetMonth() =>`, date.getMonth())
    console.log(`\ngetShortMonth() =>`, date.getShortMonth())
    console.log(`\ngetYear() =>`, date.getYear())
    console.log(`\ngetShortYear() =>`, date.getShortYear())
    console.log(`\ngetWeekdayNumber() =>`, date.getWeekdayNumber())
    console.log(`\ngetWeekday() =>`, date.getWeekday())
    console.log(`\ngetDate() =>`, date.getDate())
    console.log(`\ngetExtendedDate() =>`, date.getExtendedDate())
    console.log(`\ngetExtendedWeekdayDate() =>`, date.getExtendedWeekdayDate())
    console.log(`\ngetMonthYear() =>`, date.getMonthYear())
    console.log(`\ngetShortMonthYear() =>`, date.getShortMonthYear())
    console.log(`\ngetShortMonthShortYear() =>`, date.getShortMonthShortYear())
    console.log(`\ngetMonthNumberYear() =>`, date.getMonthNumberYear())
    console.log(`\ngetHour() =>`, date.getHour())
    console.log(`\ngetMinute() =>`, date.getMinute())
    console.log(`\ngetSecond() =>`, date.getSecond())
    console.log(`\ngetTime() =>`, date.getTime())
    console.log(`\ngetHourMinute() =>`, date.getHourMinute())
    console.log(`\ngetExtendedHour() =>`, date.getExtendedHour())
    console.log(`\ngetExtendedMinute() =>`, date.getExtendedMinute())
    console.log(`\ngetExtendedSecond() =>`, date.getExtendedSecond())
    console.log(`\ngetExtendedTime() =>`, date.getExtendedTime())
    console.log(`\ngetExtendedHourMinute() =>`, date.getExtendedHourMinute())
    console.log(`\ngetDateTime() =>`, date.getDateTime())
    console.log(`\ngetExtendedDateTime() =>`, date.getExtendedDateTime())
    console.log(`\ngetExtendedCityDate() =>`, await date.getExtendedCityDate())
    console.log(`\ngetExtendedRegionDate() =>`, await date.getExtendedRegionDate())
    console.log(`\ngetExtendedShortRegionDate() =>`, await date.getExtendedShortRegionDate())
    console.log(`\ngetExtendedCountryDate() =>`, await date.getExtendedCountryDate())
    console.log(`\ngetExtendedShortCountryDate() =>`, await date.getExtendedShortCountryDate())
    console.log(`\ngetCity() =>`, await date.getCity())
    console.log(`\ngetRegion() =>`, await date.getRegion())
    console.log(`\ngetShortRegion() =>`, await date.getShortRegion())
    console.log(`\ngetCountry() =>`, await date.getCountry())
    console.log(`\ngetShortCountry() =>`, await date.getShortCountry())
    console.log(`\ngetTimezone() =>`, await date.getTimezone())
    console.log(`\ngetZipcode() =>`, await date.getZipcode())
})()


scripts/functions-old.js
const DatePtBR = require('../lib/date-pt-br.js')
const date = require('../lib/date-pt-br-old')

console.log(`\ngetDay() =>`, date.getDay())
console.log(`\ngetMonthNumber() =>`, date.getMonthNumber())
console.log(`\ngetMonth() =>`, date.getMonth())
console.log(`\ngetYear() =>`, date.getYear())
console.log(`\ngetWeekdayNumber() =>`, date.getWeekdayNumber())
console.log(`\ngetWeekday() =>`, date.getWeekday())
console.log(`\ngetDate() =>`, date.getDate())
console.log(`\ngetDate('-') =>`, date.getDate('-'))
console.log(`\ngetExtendedDate() =>`, date.getExtendedDate())
console.log(`\ngetExtendedWeekdayDate() =>`, date.getExtendedWeekdayDate())
console.log(`\ngetExtendedCityDate() =>`, date.getExtendedCityDate('São Paulo'))
console.log(`\ngetMonthYear() =>`, date.getMonthYear())
console.log(`\ngetMonthNumberYear() =>`, date.getMonthNumberYear())
console.log(`\ngetHour() =>`, date.getHour())
console.log(`\ngetMinute() =>`, date.getMinute())
console.log(`\ngetSecond() =>`, date.getSecond())
console.log(`\ngetTime() =>`, date.getTime())
console.log(`\ngetTime('-') =>`, date.getTime('-'))
console.log(`\ngetHourMinute() =>`, date.getHourMinute())
console.log(`\ngetHourMinute('-') =>`, date.getHourMinute('-'))
console.log(`\ngetExtendedHour() =>`, date.getExtendedHour())
console.log(`\ngetExtendedMinute() =>`, date.getExtendedMinute())
console.log(`\ngetExtendedSecond() =>`, date.getExtendedSecond())
console.log(`\ngetExtendedTime() =>`, date.getExtendedTime())
console.log(`\ngetExtendedTime(', ') =>`, date.getExtendedTime(', '))
console.log(`\ngetExtendedHourMinute() =>`, date.getExtendedHourMinute())
console.log(`\ngetExtendedHourMinute(', ') =>`, date.getExtendedHourMinute(', '))
console.log(`\ngetDateTime() =>`, date.getDateTime())
console.log(`\ngetDateTime('-', ' | ', ':') =>`, date.getDateTime('-', ' | ', ':'))
console.log(`\ntoPtBr('2015-12-19T10:30:00') =>`, date.toPtBR('2015-12-19T10:30:00'))
console.log(`\ntoPtBr('2015-12-19T10:30:00', '-') =>`, date.toPtBR('2015-12-19T10:30:00', '-'))


lib/date-pt-br.js
const http = require('http')

class DatePtBR {
    constructor() {
        // Date and Time variables
        this.day = ''
        this.monthNumber = ''
        this.month = ''
        this.year = ''
        this.weekdayNumber = ''
        this.weekday = ''
        this.hour = ''
        this.minute = ''
        this.second = ''
        this.descHour = ''
        this.descMinute = ''
        this.descSecond = ''
        this.date = ''

        // Locale variables
        this.city = ''
        this.region = ''
        this.shortRegion = ''
        this.country = ''
        this.shortCountry = ''
        this.timezone = ''
        this.zipcode = ''

        // Date validations variables
        this.isDateValid = true
        this.invalidDate = 'Invalid Date'

        // Config variables
        this.setDefaultConfig()
    }

    // Set default config variables
    setDefaultConfig() {
        this.delimiterDate = '/'
        this.delimiterTime = ':'
        this.sepDateTime = ', '
        this.sepDate = ' de '
        this.sepTime = ' e '
    }

    // Alter config variables
    config(newCfg) {
        this.delimiterDate = newCfg.delimiterDate || '/'
        this.delimiterTime = newCfg.delimiterTime || ':'
        this.sepDateTime = newCfg.sepDateTime || ', '
        this.sepDate = newCfg.sepDate || ' de '
        this.sepTime = newCfg.sepTime || ' e '
    }

    // Main function
    dateNow(date) {
        try {
            this.isDateValid = true
            this.date = new Date(date)

            if (this.date == 'Invalid Date') throw new Error(`${date} is not a valid date`)

            this.day = this.date.getDate()
            this.monthNumber = this.date.getMonth() + 1
            this.year = this.date.getFullYear()
            this.weekdayNumber = this.date.getDay() + 1
            this.hour = this.date.getHours()
            this.minute = this.date.getMinutes()
            this.second = this.date.getSeconds()

            if (this.day.toString().length === 1) this.day = '0' + this.day

            if (this.monthNumber.toString().length === 1) this.monthNumber = '0' + this.monthNumber

            if (this.hour.toString().length === 1) this.hour = '0' + this.hour

            if (this.minute.toString().length === 1) this.minute = '0' + this.minute

            if (this.second.toString().length === 1) this.second = '0' + this.second

            if (this.hour !== '01') this.descHour = 'horas'
            else this.descHour = 'hora'

            if (this.minute !== '01') this.descMinute = 'minutos'
            else this.descMinute = 'minuto'

            if (this.second !== '01') this.descSecond = 'segundos'
            else this.descSecond = 'segundo'

            switch (this.weekdayNumber) {
                case 1:
                    this.weekday = 'Domingo'
                    break
                case 2:
                    this.weekday = 'Segunda-Feira'
                    break
                case 3:
                    this.weekday = 'Terça-Feira'
                    break
                case 4:
                    this.weekday = 'Quarta-Feira'
                    break
                case 5:
                    this.weekday = 'Quinta-Feira'
                    break
                case 6:
                    this.weekday = 'Sexta-Feira'
                    break
                case 7:
                    this.weekday = 'Sábado'
                    break
            }

            switch (String(this.monthNumber)) {
                case '01':
                    this.month = 'Janeiro'
                    break
                case '02':
                    this.month = 'Fevereiro'
                    break
                case '03':
                    this.month = 'Março'
                    break
                case '04':
                    this.month = 'Abril'
                    break
                case '05':
                    this.month = 'Maio'
                    break
                case '06':
                    this.month = 'Junho'
                    break
                case '07':
                    this.month = 'Julho'
                    break
                case '08':
                    this.month = 'Agosto'
                    break
                case '09':
                    this.month = 'Setembro'
                    break
                case '10':
                    this.month = 'Outubro'
                    break
                case '11':
                    this.month = 'Novembro'
                    break
                case '12':
                    this.month = 'Dezembro'
                    break
            }
        } catch (err) {
            this.isDateValid = false
        }
    }

    // Date Functions
    getDay(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? this.day : this.invalidDate
    }

    getMonthNumber(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? this.monthNumber : this.invalidDate
    }

    getMonth(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? this.month : this.invalidDate
    }

    getShortMonth(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? String(this.month).substr(0, 3) : this.invalidDate
    }

    getYear(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? this.year : this.invalidDate
    }

    getShortYear(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? String(this.year).substr(2, 4) : this.invalidDate
    }

    getWeekdayNumber(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? this.weekdayNumber : this.invalidDate
    }

    getWeekday(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? this.weekday : this.invalidDate
    }

    getDate(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.day}${this.delimiterDate}${this.monthNumber}${this.delimiterDate}${this.year}`
            : this.invalidDate
    }

    getExtendedDate(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.day}${this.sepDate}${this.month}${this.sepDate}${this.year}`
            : this.invalidDate
    }

    getExtendedWeekdayDate(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.weekday}, ${this.day}${this.sepDate}${this.month}${this.sepDate}${this.year}`
            : this.invalidDate
    }

    getMonthYear(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.month}${this.delimiterDate}${this.year}`
            : this.invalidDate
    }

    getShortMonthYear(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${String(this.month).substr(0, 3)}${this.delimiterDate}${this.year}`
            : this.invalidDate
    }

    getShortMonthShortYear(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${String(this.month).substr(0, 3)}${this.delimiterDate}${String(this.year).substr(
                  2,
                  4
              )}`
            : this.invalidDate
    }

    getMonthNumberYear(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.monthNumber}${this.delimiterDate}${this.year}`
            : this.invalidDate
    }

    getHour(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? this.hour : this.invalidDate
    }

    getExtendedHour(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? `${this.hour} ${this.descHour}` : this.invalidDate
    }

    getMinute(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? this.minute : this.invalidDate
    }

    getExtendedMinute(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? `${this.minute} ${this.descMinute}` : this.invalidDate
    }

    getSecond(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? this.second : this.invalidDate
    }

    getExtendedSecond(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid ? `${this.second} ${this.descSecond}` : this.invalidDate
    }

    getTime(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.hour}${this.delimiterTime}${this.minute}${this.delimiterTime}${this.second}`
            : this.invalidDate
    }

    getHourMinute(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.hour}${this.delimiterTime}${this.minute}`
            : this.invalidDate
    }

    getExtendedTime(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.hour} ${this.descHour}${this.sepTime}${this.minute} ${this.descMinute}${
                  this.sepTime
              }${this.second} ${this.descSecond}`
            : this.invalidDate
    }

    getExtendedHourMinute(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.hour} ${this.descHour}${this.sepTime}${this.minute} ${this.descMinute}`
            : this.invalidDate
    }

    getDateTime(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.day}${this.delimiterDate}${this.monthNumber}${this.delimiterDate}${
                  this.year
              }${this.sepDateTime}${this.hour}${this.delimiterTime}${this.minute}${
                  this.delimiterTime
              }${this.second}`
            : this.invalidDate
    }

    getExtendedDateTime(date = new Date()) {
        this.dateNow(date)
        return this.isDateValid
            ? `${this.day}${this.sepDate}${this.month}${this.sepDate}${this.year}${
                  this.sepDateTime
              }${this.hour} ${this.descHour}${this.sepTime}${this.minute} ${this.descMinute}${
                  this.sepTime
              }${this.second} ${this.descSecond}`
            : this.invalidDate
    }

    async getExtendedCityDate(date = new Date()) {
        this.dateNow(date)
        await this.getCity()
        return this.isDateValid
            ? `${this.city}, ${this.day}${this.sepDate}${this.month}${this.sepDate}${this.year}`
            : this.invalidDate
    }

    async getExtendedRegionDate(date = new Date()) {
        this.dateNow(date)
        await this.getRegion()
        return this.isDateValid
            ? `${this.region}, ${this.day}${this.sepDate}${this.month}${this.sepDate}${this.year}`
            : this.invalidDate
    }

    async getExtendedShortRegionDate(date = new Date()) {
        this.dateNow(date)
        await this.getShortRegion()
        return this.isDateValid
            ? `${this.shortRegion}, ${this.day}${this.sepDate}${this.month}${this.sepDate}${
                  this.year
              }`
            : this.invalidDate
    }

    async getExtendedCountryDate(date = new Date()) {
        this.dateNow(date)
        await this.getCountry()
        return this.isDateValid
            ? `${this.country}, ${this.day}${this.sepDate}${this.month}${this.sepDate}${this.year}`
            : this.invalidDate
    }

    async getExtendedShortCountryDate(date = new Date()) {
        this.dateNow(date)
        await this.getShortCountry()
        return this.isDateValid
            ? `${this.shortCountry}, ${this.day}${this.sepDate}${this.month}${this.sepDate}${
                  this.year
              }`
            : this.invalidDate
    }

    async getCity() {
        if (this.city === '') await this.getLocaleInfo()
        return this.city
    }

    async getRegion() {
        if (this.region === '') await this.getLocaleInfo()
        return this.region
    }

    async getShortRegion() {
        if (this.shortRegion === '') await this.getLocaleInfo()
        return this.shortRegion
    }

    async getCountry() {
        if (this.country === '') await this.getLocaleInfo()
        return this.country
    }

    async getShortCountry() {
        if (this.shortCountry === '') await this.getLocaleInfo()
        return this.shortCountry
    }

    async getTimezone() {
        if (this.timezone === '') await this.getLocaleInfo()
        return this.timezone
    }

    async getZipcode() {
        if (this.zipcode === '') await this.getLocaleInfo()
        return this.zipcode
    }

    // Get locale informations
    async getLocaleInfo() {
        return new Promise(async (resolve, reject) => {
            let req = await http.get('http://ip-api.com/json', res => {
                let data = ''
                let jsonData = ''

                res.on('data', stream => {
                    data += stream
                })
                res.on('end', () => {
                    jsonData = JSON.parse(data)

                    this.city = jsonData.city
                    this.region = jsonData.regionName
                    this.shortRegion = jsonData.region
                    this.country = jsonData.country
                    this.shortCountry = jsonData.countryCode
                    this.timezone = jsonData.timezone
                    this.zipcode = jsonData.zip

                    resolve()
                })
            })

            req.on('error', e => {
                console.log(e.message)
                reject(e.message)
            })
        })
    }
}

module.exports = DatePtBR
